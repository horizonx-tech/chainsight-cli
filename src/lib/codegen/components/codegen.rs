use anyhow::Result;
use ethabi::Contract;

use crate::{
    lib::codegen::canisters::snapshot_indexer_https::JsonTypeGenStrategyImpl,
    types::{ComponentType, Network},
};

use super::{
    algorithm_indexer::{AlgorithmIndexerCodeGenerator, AlgorithmIndexerComponentManifest},
    algorithm_lens::{AlgorithmLensCodeGenerator, AlgorithmLensComponentManifest},
    common::{ComponentManifest, GeneratedCodes},
    event_indexer::{EventIndexerCodeGenerator, EventIndexerComponentManifest},
    relayer::{RelayerCodeGenerator, RelayerComponentManifest},
    snapshot_indexer_evm::{SnapshotIndexerEVMComponentManifest, SnapshotIndexerEvmCodeGenerator},
    snapshot_indexer_https::{
        SnapshotIndesxerHTTPSCodeGenerator, SnapshotIndexerHTTPSComponentManifest,
    },
    snapshot_indexer_icp::{SnapshotIndexerICPCodeGenerator, SnapshotIndexerICPComponentManifest},
};

pub trait CodeGenerator {
    fn generate_code(&self, interface_contract: Option<Contract>) -> Result<GeneratedCodes>;
    fn generate_scripts(&self, network: Network) -> anyhow::Result<String>;
    fn generate_user_impl_template(&self) -> anyhow::Result<GeneratedCodes>;
    fn manifest(&self) -> Box<dyn ComponentManifest>;
    fn generate_component_setup_args(&self, network: &Network) -> anyhow::Result<Option<Vec<u8>>>;
}

pub fn generator(
    component_type: ComponentType,
    component_path: &str,
    id: &str,
) -> Result<Box<dyn CodeGenerator>> {
    match component_type {
        ComponentType::EventIndexer => Ok(Box::new(EventIndexerCodeGenerator::new(
            EventIndexerComponentManifest::load_with_id(component_path, id)?,
        ))),
        ComponentType::AlgorithmIndexer => Ok(Box::new(AlgorithmIndexerCodeGenerator::new(
            AlgorithmIndexerComponentManifest::load_with_id(component_path, id)?,
        ))),
        ComponentType::SnapshotIndexerICP => Ok(Box::new(SnapshotIndexerICPCodeGenerator::new(
            SnapshotIndexerICPComponentManifest::load_with_id(component_path, id)?,
        ))),
        ComponentType::SnapshotIndexerEVM => Ok(Box::new(SnapshotIndexerEvmCodeGenerator::new(
            SnapshotIndexerEVMComponentManifest::load_with_id(component_path, id)?,
        ))),
        ComponentType::Relayer => Ok(Box::new(RelayerCodeGenerator::new(
            RelayerComponentManifest::load_with_id(component_path, id)?,
        ))),
        ComponentType::AlgorithmLens => Ok(Box::new(AlgorithmLensCodeGenerator::new(
            AlgorithmLensComponentManifest::load_with_id(component_path, id)?,
        ))),
        ComponentType::SnapshotIndexerHTTPS => {
            Ok(Box::new(SnapshotIndesxerHTTPSCodeGenerator::new(
                SnapshotIndexerHTTPSComponentManifest::load_with_id(component_path, id)?,
                Box::new(JsonTypeGenStrategyImpl),
            )))
        }
    }
}
