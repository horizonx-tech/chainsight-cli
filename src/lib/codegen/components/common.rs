use std::{fs::OpenOptions, path::Path, io::Read};

use proc_macro2::TokenStream;
use serde::{Deserialize, Serialize};

use crate::types::ComponentType;

#[derive(Deserialize, Serialize, Clone, Copy, Debug, PartialEq, clap::ValueEnum)]
pub enum DatasourceType {
    #[serde(rename = "canister")]
    Canister,
    #[serde(rename = "contract")]
    Contract,
}
#[derive(Deserialize, Serialize, Clone, Copy, Debug, PartialEq, clap::ValueEnum)]
pub enum DestinactionType {
    #[serde(rename = "uint256")]
    Uint256Oracle,
    #[serde(rename = "string")]
    StringOracle,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct Datasource {
    #[serde(rename = "type")]
    pub type_: DatasourceType,
    pub id: String,
    pub method: DatasourceMethod
}
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct DatasourceMethod {
    pub identifier: String,
    pub interface: Option<String>,
    pub args: Vec<serde_yaml::Value>,
}
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct DatasourceMethodCustomStruct {
    pub name: String,
    pub fields: Vec<DatasourceMethodCustomStructField>
}
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct DatasourceMethodCustomStructField {
    pub name: String,
    #[serde(rename = "type")]
    pub type_: String
}
#[derive(Clone, Debug, Deserialize, Serialize)]

pub struct DatasourceMethodCustomType {
    pub name: String,
    pub types: Vec<String>
}

impl Datasource {
    // temp: use Default trait
    pub fn default_contract() -> Self {
        Self {
            type_: DatasourceType::Contract,
            id: "0000000000000000000000000000000000000000".to_string(), // temp
            method: DatasourceMethod {
                identifier: "totalSupply():(uint256)".to_string(),
                interface: Some("ERC20.json".to_string()),
                args: vec![],
            },
        }
    }
    pub fn new_contract(
        identifier: String,
        interface: Option<String>,
    ) -> Self {
        Self {
            type_: DatasourceType::Contract,
            id: "0000000000000000000000000000000000000000".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface,
                args: vec![],
            },
        }
    }
    
    // temp: use Default trait
    pub fn default_canister(ident_with_ts: bool) -> Self {
        let identifier = if ident_with_ts {
            "get_last_snapshot : () -> (record { value : text; timestamp : nat64 })"
        } else {
            "get_last_snapshot_value : () -> (text)"
        }.to_string();

        Self {
            type_: DatasourceType::Canister,
            id: "xxxxx-xxxxx-xxxxx-xxxxx-xxx".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface: None,
                args: vec![],
            },
        }
    }

    pub fn new_canister(
        identifier: String,
        interface: Option<String>,
    ) -> Self {
        Self {
            type_: DatasourceType::Canister,
            id: "xxxxx-xxxxx-xxxxx-xxxxx-xxx".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface,
                args: vec![],
            },
        }
    }
}

/// Determine indexer type from manifest
pub trait ComponentManifest: std::fmt::Debug {
    fn load(path: &str) -> anyhow::Result<Self> where Self: Sized;
    fn to_str_as_yaml(&self) -> anyhow::Result<String> where Self: Sized;
    fn validate_manifest(&self) -> anyhow::Result<()>;
    fn generate_codes(&self, interface_contract: Option<ethabi::Contract>) -> anyhow::Result<TokenStream>;

    fn component_type(&self) -> ComponentType;
    fn label(&self) -> &str;
    fn destination_type(&self) -> Option<DestinactionType>;
    fn required_interface(&self) -> Option<String>;
}

#[derive(Deserialize)]
pub struct ComponentTypeInManifest {
    #[serde(rename = "type")]
    pub type_: ComponentType,
}
impl ComponentTypeInManifest {
    pub fn determine_type(component_manifest_path: &str) -> anyhow::Result<ComponentType> {
        let mut file = OpenOptions::new()
            .read(true)
            .open(&Path::new(component_manifest_path))?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        let data: Self = serde_yaml::from_str(&contents)?;
        Ok(data.type_)
    }
}
