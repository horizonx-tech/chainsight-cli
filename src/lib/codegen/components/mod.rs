pub mod algorithm_indexer;
pub mod algorithm_lens;
pub mod common;
pub mod event_indexer;
pub mod relayer;
pub mod snapshot;
pub mod snapshot_json_rpc;
