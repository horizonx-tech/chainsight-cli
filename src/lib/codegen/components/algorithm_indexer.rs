use std::collections::HashMap;

use chainsight_cdk::{
    config::components::{AlgorithmIndexerConfig, AlgorithmInputType, CommonConfig},
    initializer::CycleManagements,
};
use serde::{Deserialize, Serialize};

use crate::{
    lib::codegen::{canisters, scripts},
    types::{ComponentType, Network},
    utils::serializer::ordered_map,
};

use super::common::{
    custom_tags_interval_sec, ComponentManifest, ComponentMetadata, CycleManagementsManifest,
    GeneratedCodes, SourceType, Sources,
};

/// Component Manifest: Algorithm Indexer
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct AlgorithmIndexerComponentManifest {
    #[serde(skip_serializing)]
    pub id: Option<String>,
    pub version: String,
    pub metadata: ComponentMetadata,
    pub datasource: AlgorithmIndexerDatasource,
    pub output: Vec<AlgorithmIndexerOutput>,
    pub interval: u32,
    pub cycles: Option<CycleManagementsManifest>,
}

impl From<AlgorithmIndexerComponentManifest>
    for chainsight_cdk::config::components::AlgorithmIndexerConfig
{
    fn from(val: AlgorithmIndexerComponentManifest) -> Self {
        AlgorithmIndexerConfig {
            common: CommonConfig {
                canister_name: val.id.unwrap(),
            },
            indexing: chainsight_cdk::indexer::IndexingConfig {
                start_from: val.datasource.from,
                chunk_size: None,
            },
            input: chainsight_cdk::config::components::AlgorithmIndexerInput {
                method_name: val.datasource.method,
                response_type: val.datasource.input.name,
                source_type: val.datasource.source_type,
            },
        }
    }
}

impl AlgorithmIndexerComponentManifest {
    pub fn new(
        id: &str,
        label: &str,
        description: &str,
        version: &str,
        datasource: AlgorithmIndexerDatasource,
        output: Vec<AlgorithmIndexerOutput>,
        interval: u32,
    ) -> Self {
        Self {
            id: Some(id.to_owned()),
            version: version.to_owned(),
            metadata: ComponentMetadata {
                label: label.to_owned(),
                type_: ComponentType::AlgorithmIndexer,
                description: description.to_owned(),
                tags: Some(vec!["Ethereum".to_string(), "Account".to_string()]),
            },
            datasource,
            output,
            interval,
            cycles: None,
        }
    }
}
impl ComponentManifest for AlgorithmIndexerComponentManifest {
    fn load_with_id(path: &str, id: &str) -> anyhow::Result<Self> {
        let manifest = Self::load(path)?;
        Ok(Self {
            id: Some(id.to_owned()),
            ..manifest
        })
    }

    fn to_str_as_yaml(&self) -> anyhow::Result<String> {
        let yaml = serde_yaml::to_string(&self)?;
        Ok(self.yaml_str_with_configs(yaml, "algorithm_indexer".to_string()))
    }

    fn validate_manifest(&self) -> anyhow::Result<()> {
        canisters::algorithm_indexer::validate_manifest(self)
    }

    fn generate_codes(
        &self,
        _interface_contract: Option<ethabi::Contract>,
    ) -> anyhow::Result<GeneratedCodes> {
        Ok(GeneratedCodes {
            lib: canisters::algorithm_indexer::generate_codes(self)?,
            types: None,
        })
    }

    fn generate_scripts(&self, network: Network) -> anyhow::Result<String> {
        scripts::algorithm_indexer::generate_scripts(self, network)
    }

    fn component_type(&self) -> ComponentType {
        ComponentType::AlgorithmIndexer
    }

    fn id(&self) -> Option<String> {
        self.id.clone()
    }

    fn metadata(&self) -> &ComponentMetadata {
        &self.metadata
    }

    fn destination_type(&self) -> Option<super::common::DestinationType> {
        None
    }

    fn required_interface(&self) -> Option<String> {
        None
    }

    fn generate_user_impl_template(&self) -> anyhow::Result<GeneratedCodes> {
        Ok(GeneratedCodes {
            lib: canisters::algorithm_indexer::generate_app(self)?,
            types: None,
        })
    }

    fn get_sources(&self) -> Sources {
        Sources {
            source_type: SourceType::Chainsight,
            source: self.datasource.clone().principal,
            attributes: HashMap::new(),
        }
    }
    fn custom_tags(&self) -> HashMap<String, String> {
        let mut res = HashMap::new();
        let (interval_key, interval_val) = custom_tags_interval_sec(self.interval);
        res.insert(interval_key, interval_val);
        res
    }
    fn cycle_managements(&self) -> CycleManagements {
        self.cycles.clone().unwrap_or_default().into()
    }
}
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub enum InputType {
    U128,
    U64,
    U32,
    U16,
    U8,
    String,
}
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]

pub struct InputStruct {
    pub name: String,
    #[serde(serialize_with = "ordered_map")]
    pub fields: HashMap<String, String>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct AlgorithmIndexerOutput {
    pub name: String,
    #[serde(serialize_with = "ordered_map")]
    pub fields: HashMap<String, String>,
    pub output_type: AlgorithmOutputType,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub enum AlgorithmOutputType {
    #[serde(rename = "key_values")]
    KeyValues,
    #[serde(rename = "key_value")]
    KeyValue,
}

impl Default for AlgorithmIndexerOutput {
    fn default() -> Self {
        let mut sample_fields = HashMap::new();
        sample_fields.insert("address".to_string(), "String".to_string());
        Self {
            name: "Account".to_string(),
            fields: sample_fields,
            output_type: AlgorithmOutputType::KeyValue,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct AlgorithmIndexerDatasource {
    pub principal: String,
    pub input: InputStruct,
    pub from: u64,
    pub method: String,
    pub source_type: AlgorithmInputType,
}

impl Default for AlgorithmIndexerDatasource {
    fn default() -> Self {
        let mut sample_fields = HashMap::new();
        sample_fields.insert("from".to_string(), "String".to_string());
        sample_fields.insert("to".to_string(), "String".to_string());
        sample_fields.insert("value".to_string(), "String".to_string());

        Self {
            principal: "be2us-64aaa-aaaaa-qaabq-cai".to_string(),
            input: InputStruct {
                name: "Transfer".to_string(),
                fields: sample_fields,
            },
            source_type: AlgorithmInputType::EventIndexer,
            method: "proxy_call".to_string(),
            from: 17660942,
        }
    }
}

#[cfg(test)]
mod tests {

    use insta::assert_display_snapshot;
    use jsonschema::JSONSchema;

    use crate::lib::test_utils::SrcString;

    use super::*;

    #[test]
    fn test_to_manifest_struct() {
        let yaml = r#"
version: v1
metadata:
    label: sample_algorithm_indexer
    type: algorithm_indexer
    description: Description
    tags: 
    - Ethereum
    - Account
datasource:
    principal: ahw5u-keaaa-aaaaa-qaaha-cai
    from: 17660942
    input:
        name: Transfer
        fields:
            from: String
            to: String
            value: String
    method: proxy_call
    source_type: event_indexer
output:
- name: SampleOutput
  output_type: key_value
  fields:
    result: String
    value: String
interval: 3600
"#;

        let result = serde_yaml::from_str::<AlgorithmIndexerComponentManifest>(yaml);
        let component = result.unwrap();

        assert_eq!(
            component,
            AlgorithmIndexerComponentManifest {
                id: None,
                version: "v1".to_string(),
                metadata: ComponentMetadata {
                    label: "sample_algorithm_indexer".to_string(),
                    type_: ComponentType::AlgorithmIndexer,
                    description: "Description".to_string(),
                    tags: Some(vec!["Ethereum".to_string(), "Account".to_string()])
                },
                datasource: AlgorithmIndexerDatasource {
                    input: InputStruct {
                        name: "Transfer".to_string(),
                        fields: HashMap::from([
                            ("from".to_string(), "String".to_string()),
                            ("to".to_string(), "String".to_string()),
                            ("value".to_string(), "String".to_string()),
                        ])
                    },
                    principal: "ahw5u-keaaa-aaaaa-qaaha-cai".to_string(),
                    from: 17660942,
                    method: "proxy_call".to_string(),
                    source_type: AlgorithmInputType::EventIndexer
                },
                output: vec!(AlgorithmIndexerOutput {
                    name: "SampleOutput".to_string(),
                    fields: HashMap::from([
                        ("result".to_string(), "String".to_string()),
                        ("value".to_string(), "String".to_string()),
                    ]),
                    output_type: AlgorithmOutputType::KeyValue
                }),
                interval: 3600,
                cycles: None,
            }
        );
        let schema = serde_json::from_str(include_str!(
            "../../../../resources/schema/algorithm_indexer.json"
        ))
        .expect("Invalid json");
        let instance = serde_yaml::from_str(yaml).expect("Invalid yaml");
        let compiled = JSONSchema::compile(&schema).expect("Invalid schema");
        let result = compiled.validate(&instance);
        assert!(result.is_ok());
    }

    #[test]
    fn test_snapshot_outputs() {
        let id = "sample_algorithm_indexer".to_string();
        let manifest = AlgorithmIndexerComponentManifest {
            id: Some(id.clone()),
            version: "v1".to_string(),
            metadata: ComponentMetadata {
                label: "Sample Algorithm Indexer".to_string(),
                type_: ComponentType::AlgorithmIndexer,
                description: "Description".to_string(),
                tags: Some(vec!["Ethereum".to_string(), "Account".to_string()]),
            },
            datasource: AlgorithmIndexerDatasource {
                input: InputStruct {
                    name: "Transfer".to_string(),
                    fields: HashMap::from([
                        ("from".to_string(), "String".to_string()),
                        ("to".to_string(), "String".to_string()),
                        ("value".to_string(), "String".to_string()),
                    ]),
                },
                principal: "ahw5u-keaaa-aaaaa-qaaha-cai".to_string(),
                from: 17660942,
                method: "proxy_call".to_string(),
                source_type: AlgorithmInputType::EventIndexer,
            },
            output: vec![AlgorithmIndexerOutput {
                name: "SampleOutput".to_string(),
                fields: HashMap::from([
                    ("result".to_string(), "String".to_string()),
                    ("value".to_string(), "String".to_string()),
                ]),
                output_type: AlgorithmOutputType::KeyValue,
            }],
            interval: 3600,
            cycles: None,
        };

        let snap_prefix = "snapshot__algorithm_indexer";
        let generated_codes = manifest.generate_codes(Option::None).unwrap();
        assert_display_snapshot!(
            format!("{}__canisters_lib", &snap_prefix),
            SrcString::from(generated_codes.lib)
        );
        assert!(generated_codes.types.is_none());

        let generated_user_impl_template = manifest.generate_user_impl_template().unwrap();
        assert_display_snapshot!(
            format!("{}__logics_lib", &snap_prefix),
            SrcString::from(generated_user_impl_template.lib)
        );
        assert!(generated_user_impl_template.types.is_none());

        assert_display_snapshot!(
            format!("{}__scripts", &snap_prefix),
            &manifest.generate_scripts(Network::Local).unwrap()
        );
    }
}
