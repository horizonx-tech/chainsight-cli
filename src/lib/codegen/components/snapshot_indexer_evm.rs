use std::collections::HashMap;

use chainsight_cdk::config::components::{CommonConfig, LensTargets};
use serde::{Deserialize, Serialize};
use serde_json::json;

use crate::{
    lib::codegen::{canisters, components::common::SourceType, scripts},
    types::{ComponentType, Network},
};

use super::common::{
    custom_tags_interval_sec, ComponentManifest, ComponentMetadata, Datasource, DestinationType,
    GeneratedCodes, SnapshotStorage, Sources, DEFAULT_MONITOR_DURATION_SECS,
};

/// Component Manifest: Snapshot Indexer EVM
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct SnapshotIndexerEVMComponentManifest {
    #[serde(skip_serializing)]
    pub id: Option<String>,
    pub version: String,
    pub metadata: ComponentMetadata,
    pub datasource: Datasource,
    pub storage: SnapshotStorage,
    pub interval: u32,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct LensTarget {
    pub identifiers: Vec<String>,
}

impl SnapshotIndexerEVMComponentManifest {
    pub fn new(
        id: &str,
        label: &str,
        description: &str,
        version: &str,
        datasource: Datasource,
        storage: SnapshotStorage,
        interval: u32,
    ) -> Self {
        Self {
            id: Some(id.to_owned()),
            version: version.to_owned(),
            metadata: ComponentMetadata {
                label: label.to_owned(),
                type_: ComponentType::SnapshotIndexerEVM,
                description: description.to_owned(),
                tags: Some(vec![
                    "ERC-20".to_string(),
                    "Ethereum".to_string(),
                    "DAI".to_string(),
                ]),
            },
            datasource,
            storage,
            interval,
        }
    }
}
impl From<SnapshotIndexerEVMComponentManifest>
    for chainsight_cdk::config::components::SnapshotIndexerEVMConfig
{
    fn from(val: SnapshotIndexerEVMComponentManifest) -> Self {
        let SnapshotIndexerEVMComponentManifest { id, datasource, .. } = val;
        Self {
            common: CommonConfig {
                canister_name: id.clone().unwrap(),
                monitor_duration: DEFAULT_MONITOR_DURATION_SECS,
            },
            method_identifier: datasource.method.identifier,
            method_args: datasource
                .method
                .args
                .into_iter()
                .map(yaml_to_json)
                .collect(),
            abi_file_path: format!("./__interfaces/{}", datasource.method.interface.unwrap()),
        }
    }
}
impl ComponentManifest for SnapshotIndexerEVMComponentManifest {
    fn load_with_id(path: &str, id: &str) -> anyhow::Result<Self> {
        let manifest = Self::load(path)?;
        Ok(Self {
            id: Some(id.to_owned()),
            ..manifest
        })
    }

    fn to_str_as_yaml(&self) -> anyhow::Result<String> {
        let yaml = serde_yaml::to_string(&self)?;
        Ok(self.yaml_str_with_configs(yaml, "snapshot_indexer_evm".to_string()))
    }

    fn validate_manifest(&self) -> anyhow::Result<()> {
        canisters::snapshot_indexer_evm::validate_manifest(self)
    }

    fn generate_codes(
        &self,
        _interface_contract: Option<ethabi::Contract>,
    ) -> anyhow::Result<GeneratedCodes> {
        Ok(GeneratedCodes {
            lib: canisters::snapshot_indexer_evm::generate_codes(self)?,
            types: None,
        })
    }

    fn generate_scripts(&self, network: Network) -> anyhow::Result<String> {
        scripts::snapshot_indexer_evm::generate_scripts(self, network)
    }

    fn component_type(&self) -> ComponentType {
        ComponentType::SnapshotIndexerEVM
    }

    fn id(&self) -> Option<String> {
        self.id.clone()
    }

    fn metadata(&self) -> &ComponentMetadata {
        &self.metadata
    }

    fn destination_type(&self) -> Option<DestinationType> {
        None
    }

    fn required_interface(&self) -> Option<String> {
        self.datasource.method.interface.clone()
    }

    fn generate_user_impl_template(&self) -> anyhow::Result<GeneratedCodes> {
        Ok(GeneratedCodes {
            lib: canisters::snapshot_indexer_evm::generate_app(self)?,
            types: None,
        })
    }

    fn get_sources(&self) -> Sources {
        let mut attr = HashMap::new();
        let mut method_identifier = self.datasource.clone().method.identifier;
        if method_identifier.contains(':') {
            method_identifier = method_identifier.split(':').collect::<Vec<&str>>()[0]
                .to_string()
                .replace(' ', "")
                .replace("()", "");
        }

        attr.insert("function_name".to_string(), json!(method_identifier));
        Sources {
            source: self.datasource.location.id.clone(),
            source_type: SourceType::Evm,
            attributes: attr,
        }
    }

    fn custom_tags(&self) -> HashMap<String, String> {
        let mut res = HashMap::new();
        let (interval_key, interval_val) = custom_tags_interval_sec(self.interval);
        res.insert(interval_key, interval_val);
        res
    }
}

fn yaml_to_json(yaml_value: serde_yaml::Value) -> serde_json::Value {
    // Convert serde_yaml::Value to a JSON string
    let json_str = serde_json::to_string(&yaml_value).unwrap();

    // Convert the JSON string to serde_json::Value
    serde_json::from_str(&json_str).unwrap()
}

#[cfg(test)]
mod tests {
    use std::fs::File;

    use insta::assert_display_snapshot;
    use jsonschema::JSONSchema;

    use crate::lib::{
        codegen::components::common::{DatasourceLocation, DatasourceMethod},
        test_utils::SrcString,
    };

    use super::*;

    #[test]
    fn test_to_manifest_struct() {
        let yaml = r#"
version: v1
metadata:
    label: sample_snapshot_indexer_evm
    type: snapshot_indexer_evm
    description: Description
    tags:
    - ERC-20
    - Ethereum
datasource:
    location:
        id: 6b175474e89094c44da98b954eedeac495271d0f
        args:
            network_id: 1
            rpc_url: https://mainnet.infura.io/v3/${INFURA_MAINNET_RPC_URL_KEY}
    method:
        identifier: totalSupply():(uint256)
        interface: ERC20.json
        args: []
storage:
    with_timestamp: true
interval: 3600
        "#;

        let result = serde_yaml::from_str::<SnapshotIndexerEVMComponentManifest>(yaml);
        assert!(result.is_ok());
        let component = result.unwrap();
        assert_eq!(
            component,
            SnapshotIndexerEVMComponentManifest {
                id: None,
                version: "v1".to_owned(),
                metadata: ComponentMetadata {
                    label: "sample_snapshot_indexer_evm".to_owned(),
                    type_: ComponentType::SnapshotIndexerEVM,
                    description: "Description".to_string(),
                    tags: Some(vec!["ERC-20".to_string(), "Ethereum".to_string()])
                },
                datasource: Datasource {
                    location: DatasourceLocation::new_contract(
                        "6b175474e89094c44da98b954eedeac495271d0f".to_string(),
                        1,
                        "https://mainnet.infura.io/v3/${INFURA_MAINNET_RPC_URL_KEY}".to_string(),
                    ),
                    method: DatasourceMethod {
                        identifier: "totalSupply():(uint256)".to_owned(),
                        interface: Some("ERC20.json".to_string()),
                        args: vec![]
                    }
                },
                storage: SnapshotStorage {
                    with_timestamp: true,
                },
                interval: 3600
            }
        );

        let schema = serde_json::from_str(include_str!(
            "../../../../resources/schema/snapshot_indexer_evm.json"
        ))
        .expect("Invalid json");
        let instance = serde_yaml::from_str(yaml).expect("Invalid yaml");
        let compiled = JSONSchema::compile(&schema).expect("Invalid schema");
        let result = compiled.validate(&instance);
        assert!(result.is_ok());
    }

    #[test]
    fn test_snapshot_outputs() {
        let manifest = SnapshotIndexerEVMComponentManifest {
            id: Some("sample_snapshot_indexer_evm".to_owned()),
            version: "v1".to_owned(),
            metadata: ComponentMetadata {
                label: "Sample Snapshot Indexer Evm".to_owned(),
                type_: ComponentType::SnapshotIndexerEVM,
                description: "Description".to_string(),
                tags: Some(vec!["ERC-20".to_string(), "Ethereum".to_string()]),
            },
            datasource: Datasource {
                location: DatasourceLocation::new_contract(
                    "6b175474e89094c44da98b954eedeac495271d0f".to_string(),
                    1,
                    "https://mainnet.infura.io/v3/${INFURA_MAINNET_RPC_URL_KEY}".to_string(),
                ),
                method: DatasourceMethod {
                    identifier: "totalSupply():(uint256)".to_owned(),
                    interface: Some("ERC20.json".to_string()),
                    args: vec![],
                },
            },
            storage: SnapshotStorage {
                with_timestamp: true,
            },
            interval: 3600,
        };

        let snap_prefix = "snapshot__snapshot_indexer_evm";
        let abi = File::open("resources/ERC20.json").unwrap();
        let generated_codes = manifest
            .generate_codes(Option::Some(ethabi::Contract::load(abi).unwrap()))
            .unwrap();
        assert_display_snapshot!(
            format!("{}__canisters_lib", &snap_prefix),
            SrcString::from(generated_codes.lib)
        );
        assert!(generated_codes.types.is_none());

        let generated_user_impl_template = manifest.generate_user_impl_template().unwrap();
        assert_display_snapshot!(
            format!("{}__logics_lib", &snap_prefix),
            SrcString::from(generated_user_impl_template.lib)
        );
        assert!(generated_user_impl_template.types.is_none());

        assert_display_snapshot!(
            format!("{}__scripts", &snap_prefix),
            &manifest.generate_scripts(Network::Local).unwrap()
        );
    }
}
