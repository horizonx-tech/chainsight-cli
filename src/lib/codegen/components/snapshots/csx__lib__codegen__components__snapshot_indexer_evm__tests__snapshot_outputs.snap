---
source: src/lib/codegen/components/snapshot_indexer_evm.rs
expression: "SrcString::from(&generated_codes.lib)"
---
use std::str::FromStr;
use candid::{
Decode, Encode };
use chainsight_cdk_macros::{
init_in, manage_single_state, setup_func, prepare_stable_structure, stable_memory_for_vec, StableMemoryStorable, timer_task_func, define_transform_for_web3, define_web3_ctx, chainsight_common, did_export, snapshot_web3_source };
use ic_web3_rs::types::Address;
init_in!();
chainsight_common!(3600);
define_web3_ctx!();
define_transform_for_web3!();
manage_single_state!("target_addr", String, false);
setup_func!({
target_addr:String, web3_ctx_param:chainsight_cdk::web3::Web3CtxParam
});
prepare_stable_structure!();
stable_memory_for_vec!("snapshot", Snapshot, 0, true);
timer_task_func!("set_task", "execute_task", true);
#[derive(Debug, Clone, candid::CandidType, candid::Deserialize, serde::Serialize, StableMemoryStorable)]
#[stable_mem_storable_opts(max_size = 10000, is_fixed_size = false)]
pub struct Snapshot {
pub value:SnapshotValue, pub timestamp:u64,
}

type SnapshotValue =(String);
fn _get_last_snapshot_value() -> SnapshotValue {
get_last_snapshot().value
}

fn _get_top_snapshot_values(n:u64) -> Vec<SnapshotValue>{
get_top_snapshots(n).iter().map(| s | s.value.clone()).collect()
}

fn _get_snapshot_value(idx:u64) -> SnapshotValue {
get_snapshot(idx).value
}

#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_last_snapshot_value() -> SnapshotValue {
_get_last_snapshot_value()
}

#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_top_snapshot_values(n:u64) -> Vec<SnapshotValue>{
_get_top_snapshot_values(n)
}

#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_snapshot_value(idx:u64) -> SnapshotValue {
_get_snapshot_value(idx)
}

#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_last_snapshot_value(input:std::vec::Vec<u8>) -> std::vec::Vec<u8>{
use chainsight_cdk::rpc::Receiver;
chainsight_cdk::rpc::ReceiverProviderWithoutArgs::< SnapshotValue>:: new(proxy(), _get_last_snapshot_value, ).reply(input).await
}

#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_top_snapshot_values(input:std::vec::Vec<u8>) -> std::vec::Vec<u8>{
use chainsight_cdk::rpc::Receiver;
chainsight_cdk::rpc::ReceiverProvider::< u64, Vec<SnapshotValue>>::new(proxy(), _get_top_snapshot_values, ).reply(input).await
}

#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_snapshot_value(input:std::vec::Vec<u8>) -> std::vec::Vec<u8>{
use chainsight_cdk::rpc::Receiver;
chainsight_cdk::rpc::ReceiverProvider::< u64, SnapshotValue>:: new(proxy(), _get_snapshot_value, ).reply(input).await
}

ic_solidity_bindgen::contract_abi!("./__interfaces/ERC20.json");
snapshot_web3_source!("total_supply");
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn execute_task() {
if ic_cdk::caller()!= proxy() {
panic!("Not permitted")
}

let current_ts_sec = ic_cdk::api::time() / 1000000;
let res = ERC20::new(Address::from_str(& get_target_addr()).unwrap(), & web3_ctx().unwrap()).total_supply(None).await.unwrap();
let datum = Snapshot {
value:(res.to_string()), timestamp:current_ts_sec, };
add_snapshot(datum.clone());
ic_cdk::println!("ts={

},  snapshot={
:?
}", datum.timestamp, datum.value);

}

did_export!("sample_snapshot_indexer_evm");

